R/mlClassificationNeuralNetwork.R

```diff
@@ -0,0 +1,265 @@
+  .classificationDecisionBoundaries(dataset, options, jaspResults, ready, position = 12, type = "neuralnet")
+  
+}
+
+.neuralnetClassification <- function(dataset, options, jaspResults) {
+  
+  # Import model formula from jaspResults
+  formula <- jaspResults[["formula"]]$object
+  
+  # Split the data into training and test sets
+  if (options[["holdoutData"]] == "testSetIndicator" && options[["testSetIndicatorVariable"]] != "") {
+    # Select observations according to a user-specified indicator (included when indicator = 1)
+    train.index <- which(dataset[,options[["testSetIndicatorVariable"]]] == 0)
+  } else {
+    # Sample a percentage of the total data set
+    train.index <- sample.int(nrow(dataset), size = ceiling( (1 - options[['testDataManual']]) * nrow(dataset)))
+  }
+  trainAndValid <- dataset[train.index, ]
+  
+  # Create the generated test set indicator
+  testIndicatorColumn <- rep(1, nrow(dataset))
+  testIndicatorColumn[train.index] <- 0
+  
+  if (options[["modelOpt"]] == "optimizationManual") {
+    # Just create a train and a test set (no optimization)
+    train <- trainAndValid
+    test <- dataset[-train.index, ]
+    
+    # Structure of hidden layers
+    structure <- .getNeuralNetworkStructure(options)
+    
+    p <- try({
+      nfit_test  <- neuralnet::neuralnet(formula = formula, 
+                                         data = train, 
+                                         hidden = structure,
+                                         learningrate = options[["learningRate"]],
+                                         threshold = options[["threshold"]],
+                                         stepmax = options[["stepMax"]],
+                                         rep = 1,
+                                         startweights = NULL,
+                                         algorithm = options[["algorithm"]],
+                                         err.fct = "sse",# jaspResults[["errfct"]]$object,
+                                         act.fct = jaspResults[["actfct"]]$object, 
+                                         linear.output = if (options[["actfct"]] == "linear") TRUE else FALSE)
+    })
+    if (isTryError(p)) # kind off harsh right now
+      jaspBase:::.quitAnalysis(gettext("The network did not converge."))
+    
+  } else if (options[["modelOpt"]] == "optimizationError") {
+    # Genetic optimization of network topology
+    
+    # Create a train, validation and test set (optimization)
+    valid.index             <- sample.int(nrow(trainAndValid), size = ceiling(options[['validationDataManual']] * nrow(trainAndValid)))
+    test                    <- dataset[-train.index, ]
+    valid                   <- trainAndValid[valid.index, ]
+    train                   <- trainAndValid[-valid.index, ]
+    
+    accuracyStore           <- numeric(options[["maxGen"]])
+    trainAccuracyStore      <- numeric(options[["maxGen"]])
+    
+    # For plotting
+    plot_x            <- numeric()
```

Comment:
Yeah let's not concatenate a vector in a loop. It's better and faster to instantiate a vector of the expected length and then fill up the vector.